src/audio/tdfb/tdfb.c

```diff
@@ -0,0 +1,668 @@
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	cd->tdfb_func = NULL;
+	cd->config = NULL;
+	cd->config_new = NULL;
+	cd->config_ready = false;
+	cd->fir_delay = NULL;
+	cd->fir_delay_size = 0;
+
+	/* Allocate and make a copy of the coefficients blob and reset FIR. If
+	 * the filters are configured later in run-time the size is zero.
+	 */
+	if (bs) {
+		comp_cl_info(&comp_tdfb, "tdfb_new(), allocating %d for configuration blob",
+			     bs);
+		cd->config = rballoc(0, SOF_MEM_CAPS_RAM, bs);
+		if (!cd->config) {
+			comp_cl_err(&comp_tdfb, "tdfb_new(), failed rballoc()");
+			rfree(dev);
+			rfree(cd);
+			return NULL;
+		}
+
+		ret = memcpy_s(cd->config, bs, ipc_tdfb->data, bs);
+		assert(!ret);
+		cd->config_ready = true;
+	}
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		fir_reset(&cd->fir[i]);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void tdfb_free(struct comp_dev *dev)
+{
+	struct tdfb_comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "tdfb_free()");
+
+	tdfb_free_delaylines(cd);
+	tdfb_free_parameters(&cd->config);
+	tdfb_free_parameters(&cd->config_new);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static int tdfb_cmd_get_data(struct comp_dev *dev,
+			     struct sof_ipc_ctrl_data *cdata, int max_size)
+{
+	struct tdfb_comp_data *cd = comp_get_drvdata(dev);
+	unsigned char *dst, *src;
+	size_t offset;
+	size_t bs;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "tdfb_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+
+		max_size -= sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+
+		/* Copy back to user space */
+		if (cd->config) {
+			src = (unsigned char *)cd->config;
+			dst = (unsigned char *)cdata->data->data;
+			bs = cd->config->size;
+			cdata->elems_remaining = 0;
+			offset = 0;
+			if (bs > max_size) {
+				bs = (cdata->msg_index + 1) * max_size > bs ?
+					bs - cdata->msg_index * max_size :
+					max_size;
+				offset = cdata->msg_index * max_size;
+				cdata->elems_remaining = cd->config->size -
+					offset;
+			}
+			cdata->num_elems = bs;
+			comp_info(dev, "tdfb_cmd_get_data(), blob size %zu msg index %u max size %u offset %zu",
+				  bs, cdata->msg_index, max_size, offset);
+			ret = memcpy_s(dst, ((struct sof_abi_hdr *)
+				       (cdata->data))->size, src + offset,
+				       bs);
+			assert(!ret);
+
+			cdata->data->abi = SOF_ABI_VERSION;
+			cdata->data->size = bs;
+		} else {
+			comp_err(dev, "tdfb_cmd_get_data() error: NULL cd->config");
+			ret = -EINVAL;
+		}
+		break;
+	default:
+		comp_err(dev, "tdfb_cmd_get_data() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int tdfb_cmd_set_data(struct comp_dev *dev,
+			     struct sof_ipc_ctrl_data *cdata)
+{
+	struct tdfb_comp_data *cd = comp_get_drvdata(dev);
+	unsigned char *dst, *src;
+	uint32_t offset;
+	size_t size;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
```

Comment:
Note: Once PR #3240 is available this component should use the generic functions to set/get the configuration blob.

Reply:
This is done into current PR version. Nice code lines count reduction!

Reply:
Resolved.